### PR TITLE
[MM-24945] Refactor job store to use squirrel

### LINKED ISSUE #14501: Refactor SQL queries in store/sqlstore/job_store.go to use the squirrel builder
Mattermost is replacing the SQL queries of the store layer to use the [squirrel builder](https://github.com/Masterminds/squirrel). This help wanted issue is to change the usages of raw SQL queries in the file `store/sqlstore/job_store.go`.

The expected way to implement this is to search in the `store/sqlstore/job_store.go` file for all the strings containing a SQL query and rewrite them with squirrel. To do so, we must obtain the query builder using the `getQueryBuilder()` method from the store struct, and use it to build the same query that we have in the string.

You can find more information on how to use squirrel in [the official package documentation](https://godoc.org/github.com/Masterminds/squirrel).

Example Pull Request: https://github.com/mattermost/mattermost-server/pull/14393

If you have questions about the ticket or you need any help, feel free to contact `miguel.delacruz` in https://community.mattermost.com/

---
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-24945

### PR DESCRIPTION
**Pull request by dantepippi**
**Summary**

This pull request refactors the job store to use Squirrel instead of raw sql queries.

Closes #14501

### CODE REVIEW

File: store/sqlstore/job_store.go [hunk 1]
```diff
@@ -42,16 +44,16 @@ func (jss SqlJobStore) Save(job *model.Job) (*model.Job, *model.AppError) {
...
-		"Id":             job.Id,
-		"OldStatus":      currentStatus,
-		"LastActivityAt": model.GetMillis(),
-		"Status":         job.Status,
-		"Data":           job.DataToJson(),
-		"Progress":       job.Progress,
-	}
-	sqlResult, err := jss.GetMaster().Exec(query, params)
+	sql, args, err := jss.getQueryBuilder().
+		Update("").
```
Comment (jespino) [inline store/sqlstore/job_store.go:48]: You can use here `Update("Jobs")` instead of `Udpate("").Table("Jobs")`

File: store/sqlstore/job_store.go [hunk 2]
```diff
@@ -42,16 +44,16 @@ func (jss SqlJobStore) Save(job *model.Job) (*model.Job, *model.AppError) {
...
+		Table("Jobs").
+		SetMap(sq.Eq{"LastActivityAt": model.GetMillis(), "Status": job.Status, "Data": job.DataToJson(),
```
Comment (jespino) [inline store/sqlstore/job_store.go:50]: I prefer here multiple set calls, because we already know the fields. something like:
```go
Set("LastActivityAt", model.GetMillis()).
Set("Status", job.Status).
...
```

File: store/sqlstore/job_store.go [hunk 3]
```diff
@@ -86,20 +88,21 @@ func (jss SqlJobStore) UpdateStatus(id string, status string) (*model.Job, *mode
 }
 
 func (jss SqlJobStore) UpdateStatusOptimistically(id string, currentStatus string, newStatus string) (bool, *model.AppError) {
-	var startAtClause string
+	sql := jss.getQueryBuilder().
+		Update("").
```
Comment (jespino) [inline store/sqlstore/job_store.go:92]: Same here about the table name

File: store/sqlstore/job_store.go [hunk 4]
```diff
@@ -86,20 +88,21 @@ func (jss SqlJobStore) UpdateStatus(id string, status string) (*model.Job, *mode
...
+		Table("Jobs").
+		SetMap(sq.Eq{"LastActivityAt": model.GetMillis(), "Status": newStatus}).
```
Comment (jespino) [inline store/sqlstore/job_store.go:94]: Same here about using multiple times set instead of SetMap.

File: store/sqlstore/job_store.go [hunk 5]
```diff
@@ -128,66 +136,118 @@ func (jss SqlJobStore) Get(id string) (*model.Job, *model.AppError) {
...
 		return int64(0), model.NewAppError("SqlJobStore.GetCountByStatusAndType", "store.sql_job.get_count_by_status_and_type.app_error", nil, "Status="+status+", "+err.Error(), http.StatusInternalServerError)
 	}
 	return count, nil
 }
 
 func (jss SqlJobStore) Delete(id string) (string, *model.AppError) {
-	query := "DELETE FROM Jobs WHERE Id = :Id"
-	if _, err := jss.GetMaster().Exec(query, map[string]interface{}{"Id": id}); err != nil {
+	sql, args, err := jss.getQueryBuilder().
+		Delete("").
```
Comment (jespino) [inline store/sqlstore/job_store.go:243]: You can use directly `Delete("Jobs")` instead of `Delete("").From("Jobs")`

Comment (dantepippi) [reply store/sqlstore/job_store.go]: Done.

File: store/sqlstore/job_store.go (hunk 4 above)
Comment (dantepippi) [reply store/sqlstore/job_store.go]: Done.

File: store/sqlstore/job_store.go (hunk 2 above)
Comment (dantepippi) [reply store/sqlstore/job_store.go]: Done.

File: store/sqlstore/job_store.go (hunk 1 above)
Comment (dantepippi) [reply store/sqlstore/job_store.go]: Done.

File: store/sqlstore/job_store.go (hunk 3 above)
Comment (dantepippi) [reply store/sqlstore/job_store.go]: Done.

File: store/sqlstore/job_store.go [hunk 6]
```diff
@@ -86,20 +89,21 @@ func (jss SqlJobStore) UpdateStatus(id string, status string) (*model.Job, *mode
...
-	query := "UPDATE Jobs SET " + startAtClause + " Status = :NewStatus, LastActivityAt = :LastActivityAt WHERE Id = :Id AND Status = :OldStatus"
-	params := map[string]interface{}{
-		"Id":             id,
-		"OldStatus":      currentStatus,
-		"NewStatus":      newStatus,
-		"StartAt":        model.GetMillis(),
-		"LastActivityAt": model.GetMillis(),
+	query, args, err := sql.ToSql()
+	if err != nil {
+		return false, model.NewAppError("SqlJobStore.UpdateStatus", "store.sql.build_query.app_error", nil, err.Error(), http.StatusInternalServerError)
```
Comment (mgdelacroix) [inline store/sqlstore/job_store.go:103]: nit:
```suggestion
		return false, model.NewAppError("SqlJobStore.UpdateStatusOptimistically", "store.sql.build_query.app_error", nil, err.Error(), http.StatusInternalServerError)
```

File: store/sqlstore/job_store.go [hunk 7]
```diff
@@ -86,20 +89,21 @@ func (jss SqlJobStore) UpdateStatus(id string, status string) (*model.Job, *mode
...
 	}
 
-	sqlResult, err := jss.GetMaster().Exec(query, params)
+	sqlResult, err := jss.GetMaster().Exec(query, args...)
 	if err != nil {
 		return false, model.NewAppError("SqlJobStore.UpdateStatus", "store.sql_job.update.app_error", nil, "id="+id+", "+err.Error(), http.StatusInternalServerError)
```
Comment (mgdelacroix) [inline store/sqlstore/job_store.go:108]: Although is not part of this PR, for consistency this one should be:

```suggestion
		return false, model.NewAppError("SqlJobStore.UpdateStatusOptimistically", "store.sql_job.update.app_error", nil, "id="+id+", "+err.Error(), http.StatusInternalServerError)
```

**Comment by teresa-novoa** [comment]
@jespino @mgdelacroix looks like this PR is ready to be merged, can you help? Gracias!

**Comment by lindy65** [comment]
Hi @jespino or @mgdelacroix - can you help with test steps for this PR please?  Is smoke-testing statuses correct?  I'm not certain what the "job store" entails... Thanks!
